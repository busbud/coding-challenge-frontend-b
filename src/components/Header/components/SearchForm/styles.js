import styled from "styled-components";

import { BREAKPOINTS } from "utils/constants";

export const Form = styled.div`
  background-color: var(--white);
  border-radius: 4px;
  margin-top: 16px;
  padding: 16px;

  @media screen and (min-width: ${BREAKPOINTS.tablet}) {
    align-items: center;
    display: flex;
    padding: 12px 16px;
  }
`;

export const FormControl = styled.div`
  &:not(:last-of-type) {
    margin-bottom: 16px;
  }

  @media screen and (min-width: ${BREAKPOINTS.tablet}) {
    flex: 1;
    margin-bottom: 0;

    &:not(:last-of-type) {
      margin-right: 16px;
      margin-bottom: 0;
    }
  }
`;
