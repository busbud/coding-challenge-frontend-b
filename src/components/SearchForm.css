.SearchForm .datepicker {
  padding: 0 1em; }
