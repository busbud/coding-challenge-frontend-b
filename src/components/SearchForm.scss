.SearchForm {
  .datepicker {
    padding: 0 1.75em;
  }
}