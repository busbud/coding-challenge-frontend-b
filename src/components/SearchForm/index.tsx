import * as React from 'react'
import { Trans } from 'react-i18next'

interface SearchFormProps {
  buttonClick: (ev: React.MouseEvent<HTMLButtonElement>) => void
}

interface SearchFormState {
  from: string,
  to: string,
  date: string,
  passengers: string
}

export default class SearchForm extends React.Component<SearchFormProps, SearchFormState> {
  constructor (props: SearchFormProps) {
    super(props)
    this.state = {
      from: 'New York',
      to: 'Montreal',
      date: '2019-08-02',
      passengers: '1 adult'
    }
  }

  render () {
    return (
      <section className='search'>
        <div className='search-container container'>
          <div className='row justify-content-center align-items-end'>
            <div className='col-lg-2'>
              <label htmlFor='from'>
                <Trans>Departure</Trans>
              </label>
              <input type='text' value={this.state.from} name='from' className='form-input search-from' aria-label='from' readOnly/>
            </div>
            <div className='col-lg-2'>
              <label htmlFor='to'><Trans>Arrival</Trans></label>
              <input type='text' value={this.state.to} name='to' className='form-input search-to' aria-label='to' readOnly/>
            </div>
            <div className='col-lg-3'>
              <label htmlFor='from'>Date</label>
              <input type='date' value={this.state.date} name='date' className='form-input' arial-label='date' readOnly/>
            </div>
            <div className='col-lg-2'>
              <label htmlFor='passenger'><Trans>Passengers</Trans></label>
              <input type='text' value={this.state.passengers} name='passengers' className='form-input' arial-label='passenger' readOnly/>
            </div>
            <div className='col-lg-2'>
              <button type='button' className='form-button' onClick={this.props.buttonClick}><Trans>search</Trans></button>
            </div>
          </div>
        </div>
      </section>
    )
  }
}
