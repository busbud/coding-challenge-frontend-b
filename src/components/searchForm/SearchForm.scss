.SearchForm{
    display: flex;
    align-items: center;
    justify-content: center;
    margin-top: 100px;

    .text-field {
        margin-right: 20px;
    }
}
