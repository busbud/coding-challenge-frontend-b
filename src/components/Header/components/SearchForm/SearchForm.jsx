import React, { useState } from "react";
import PropTypes from "prop-types";
import { useTranslation } from "react-i18next";

import Select from "commons/Select";
import DatePicker from "commons/DatePicker";
import Button from "commons/Button";

import { Form, FormControl } from "./styles";

const SearchForm = ({ onSearch, ...otherProps }) => {
  const { t } = useTranslation();
  const [originCity, setOriginCity] = useState("");
  const [destinationCity, setDestinationCity] = useState("");
  const [passengers, setPassengers] = React.useState("");
  const [outboundDate, setOutboundDate] = React.useState(null);

  const handleOriginCityChange = (value) => {
    setOriginCity(value);
  };

  const handleDestinationCityChange = (value) => {
    setDestinationCity(value);
  };

  const handlePassengersChange = (value) => {
    setPassengers(value);
  };

  const handleDateChange = (date, dateString) => {
    setOutboundDate(dateString);
  };

  const handleSearch = () => {
    onSearch({
      originCity,
      destinationCity,
      outboundDate,
      passengers,
    });
  };

  return (
    <Form {...otherProps}>
      <FormControl>
        <Select
          style={{ width: "100%" }}
          defaultValue={originCity}
          onChange={handleOriginCityChange}
          data-testid="origin-city-select"
        >
          <Select.Option value="">
            <em>{t("originCity")}</em>
          </Select.Option>
          <Select.Option value="f2m673">Québec</Select.Option>
        </Select>
      </FormControl>
      <FormControl>
        <Select
          style={{ width: "100%" }}
          defaultValue={destinationCity}
          onChange={handleDestinationCityChange}
          data-testid="destination-city-select"
        >
          <Select.Option value="">
            <em>{t("destinationCity")}</em>
          </Select.Option>
          <Select.Option value="f25dvk">Montréal</Select.Option>
        </Select>
      </FormControl>
      <FormControl>
        <Select
          style={{ width: "100%" }}
          defaultValue={passengers}
          onChange={handlePassengersChange}
          data-testid="passengers-select"
        >
          <Select.Option value="">
            <em>{t("passengers")}</em>
          </Select.Option>
          <Select.Option value="1">1 {t("adult")}</Select.Option>
        </Select>
      </FormControl>
      <FormControl>
        <DatePicker
          style={{ width: "100%" }}
          onChange={handleDateChange}
          data-testid="outbound-date-picker"
          placeholder="Outbound Date"
        />
      </FormControl>
      <FormControl>
        <Button
          variant="contained"
          color="primary"
          onClick={handleSearch}
          data-testid="search-button"
        >
          {t("search")}
        </Button>
      </FormControl>
    </Form>
  );
};

SearchForm.propTypes = {
  /**
   * The search event trigger.
   */
  onSearch: PropTypes.func.isRequired,
};

export default SearchForm;
