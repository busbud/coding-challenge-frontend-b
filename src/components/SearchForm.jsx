import React from "react";
import 'flatpickr/dist/themes/material_green.css'
import Flatpickr from 'react-flatpickr'
import './SearchForm.css'
import utils from "../utils/utils";

class SearchForm extends React.Component {

  date = this.props.search.date;
  origin = utils.geohashToName(this.props.search.origin);
  destination = utils.geohashToName(this.props.search.destination);

  cityInput(city, onChange) {
    return (
      <div className="field-body is-small">
        <div className="field">
          <p className="control">
            <input
              type="text"
              disabled
              className="input is-small"
              onChange={onChange}
              value={city}
            />
          </p>
        </div>
      </div>
    );
  }

  datePicker = (
    <div className="control datepicker">
      <Flatpickr
        className="input is-small" value={this.date}
        onChange={(newDate) => {
          this.date = newDate[0]
        }}
      />
    </div>
  );

  submitButton = (
    <div className="control">
      <button
        onClick={() => {
          this.props.onSearchClick(this.props.search.origin, this.props.search.destination, this.date)
        }}
        className="button is-primary is-small">Search</button>
    </div>
  );

  render() {
    return (
      <div className="SearchForm field is-horizontal">
        <div className="field-label is-small">
          <label className="label">From</label>
        </div>
        {this.cityInput(this.origin, this.props.onOriginChange)}
        <div className="field-label is-small">
          <label className="label">To</label>
        </div>
        {this.cityInput(this.destination, this.props.onDestinationChange)}

        {this.datePicker}

        {this.submitButton}
      </div>
    );
  }
};

export default SearchForm;
