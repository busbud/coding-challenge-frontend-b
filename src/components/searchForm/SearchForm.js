import React, { Component } from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import { connect } from 'react-redux';
import { wait } from '../../helpers/wait';
import { cleanStore, searchRequestSuccess } from '../../actions/search.actions';
import { URL_INIT_SEARCH, URL_POLL_SEARCH, HEARDERS } from '../../constants/applications';
import './SearchForm.scss';

export class SearchForm extends Component {
    constructor() {
        super();
        this.state = {
            departure: "New York",
            arrival: "Montreal",
            date: "2019-08-02",
            passenger: "1 Adult",
            departures: []
        }
        this.handleSearch = this.handleSearch.bind(this);
    }

    async handleSearch() {
        this.props.cleanStore();
        try {
            const response = await fetch(URL_INIT_SEARCH, HEARDERS)
            if(response.status === 200) {
                await this.fetchWithIndex(URL_POLL_SEARCH, 10);
            }
        } catch (err) {
            console.log('Init search failed');
        }
    }
    
    async fetchWithIndex(url, index){
        try {
            // const WITH_INDEX = `${url}${index}`;
            index += 10;
            const response = await fetch(url, HEARDERS);
            const data = await response.json();
            this.props.searchRequestSuccess(data)
    
            if(!data.complete){
                await wait(3000);
                fetchWithIndex(URL_POLL_SEARCH, index);
            }
    
        } catch (err) {
            console.log('Poll search failed');
        }
    }

    render() {
        return (
            <div className="SearchForm">
                <TextField
                    label="Leaving from"
                    className="text-field"
                    value={this.state.departure}
                    margin="normal"
                    variant="outlined"
                />
                <TextField
                    label="Going to"
                    className="text-field"
                    value={this.state.arrival}
                    margin="normal"
                    variant="outlined"
                />
                <TextField
                    label="Date"
                    className="text-field"
                    value={this.state.date}
                    margin="normal"
                    variant="outlined"
                />
                <TextField
                    label="passenger"
                    className="text-field"
                    value={this.state.passenger}
                    margin="normal"
                    variant="outlined"
                />
                <Button
                    variant="contained"
                    color="primary"
                    onClick={this.handleSearch}
                >
                    Search
                </Button>
            </div>
        )
    }
}

const mapDispatchToProps = {
    cleanStore,
    searchRequestSuccess
}

export default connect(undefined, mapDispatchToProps)(SearchForm);