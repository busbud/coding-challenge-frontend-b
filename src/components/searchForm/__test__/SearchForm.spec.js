import React from 'react';
import {SearchForm} from '../SearchForm';
import renderer from 'react-test-renderer';

jest.mock('@material-ui/core/TextField', () => 'TextField');
jest.mock('@material-ui/core/Paper', () => 'Button');

const props = {
    cleanStore: () => {},
    searchRequestSuccess: () => {}
};
describe('SearchForm', () => {
    it('renders correctly', () => {
        const tree = renderer
            .create(<SearchForm {...props} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
})
