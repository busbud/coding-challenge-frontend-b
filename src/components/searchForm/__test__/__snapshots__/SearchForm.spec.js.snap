// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders correctly 1`] = `
<div
  className="SearchForm"
>
  <TextField
    className="text-field"
    label="Leaving from"
    margin="normal"
    value="New York"
    variant="outlined"
  />
  <TextField
    className="text-field"
    label="Going to"
    margin="normal"
    value="Montreal"
    variant="outlined"
  />
  <TextField
    className="text-field"
    label="Date"
    margin="normal"
    value="2019-08-02"
    variant="outlined"
  />
  <TextField
    className="text-field"
    label="passenger"
    margin="normal"
    value="1 Adult"
    variant="outlined"
  />
  <button
    className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-2"
    >
      Search
    </span>
    <span
      className="MuiTouchRipple-root-30"
    />
  </button>
</div>
`;
